README.md
# flutter_advanced_progress
An advanced progress widget, that can be fully customized with bunch of properties, just try it and enjoy!

| State 1 | State 2 |
|:-:|:-:|
| ![Flutter Advanced Progress Light](./PREVIEW_LIGHT.png) | ![Flutter Advanced Progress Dark](./PREVIEW_DARK.png) |


## AdvancedProgress Parameters
|Parameter|Description|Type|Default|
|:--------|:----------|:---|:------|
|`primaryValue`|Value for primary progress.|*double*||
|`secondaryValue`|Value for secondary progress.|*double*||
|`secondaryWidth`|Secondary progress width.|*double*|10.0|
|`radius`|Total radius for whole widget.|*double*||
|`startAngle`|Progress start angle.|*double*|120.0|
|`maxDegrees`|Progress degrees from [startAngle].|*double*|300.0|
|`progressGap`|Gap between primary and secondary progress.|*double*|0.0|
|`division`|Primary progress division.|*int*|10|
|`levelAmount`|Amount of levels on primary progress.|*int*||
|`levelLowWidth`|Width of levels on primary progress.|*double*|1.0|
|`levelLowHeight`|Height of low levels on primary progress.|*double*|8.0|
|`levelHighWidth`|Height of high levels managed by [division] on primary progress.|*double*|2.0|
|`levelHighHeight`|Width of levels on primary progress.|*double*|16.0|
|`levelHighBeginEnd`|True if need to begin and end with high level.|*bool*|false|
|`primaryColor`|Primary color that used as a color for progress of first in gradient.|*Color*||
|`secondaryColor`|Secondary color that used last in gradient.|*Color*||
|`tertiaryColor`|Tertiary color that used for inactive part of progress.|*Color*||
|`child`|Child widget.|*Widget*||


CHANGELOG.md
## 1.0.0+1

* LICENSE CHANGED.

## 1.0.0

* Null-Safety implemented. 

## 0.0.3

* Bunch of parameters added.
* Examples improved.

## 0.0.2

* Documentation improvements.

## 0.0.1

* Basic functionality implemented.

example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_advanced_progress_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) => widget is Text &&
                           widget.data.startsWith('Running on:'),
      ),
      findsOneWidget,
    );
  });
}


lib/src/paint.dart
part of '../flutter_advanced_progress.dart';

/// Advanced progress painter.
class AdvancedProgressPainter extends CustomPainter {
  const AdvancedProgressPainter({
    this.primaryValue,
    this.secondaryValue,
    this.secondaryWidth,
    this.radius,
    this.startAngle,
    this.maxDegrees,
    this.progressGap,
    this.division,
    this.levelAmount,
    this.levelLowWidth,
    this.levelLowHeight,
    this.levelHighWidth,
    this.levelHighHeight,
    this.primaryColor,
    this.secondaryColor,
    this.tertiaryColor,
    this.levelHighBeginEnd,
  });

  /// Value for primary progress.
  final double? primaryValue;

  /// Value for secondary progress.
  final double? secondaryValue;

  /// Secondary progress width.
  final double? secondaryWidth;

  /// Total radius for whole widget.
  final double? radius;

  /// Progress start angle.
  final double? startAngle;

  /// Progress degrees from [startAngle].
  final double? maxDegrees;

  /// Gap between primary and secondary progress.
  final double? progressGap;

  /// Primary progress division.
  final int? division;

  /// Amount of levels on primary progress.
  final int? levelAmount;

  /// Width of levels on primary progress.
  final double? levelLowWidth;

  /// Height of low levels on primary progress.
  final double? levelLowHeight;

  /// Height of high levels managed by [division] on primary progress.
  final double? levelHighHeight;

  /// Width of levels on primary progress.
  final double? levelHighWidth;

  /// True if need to begin and end with high level.
  final bool? levelHighBeginEnd;

  /// Primary color that used as a color for progress of first in gradient.
  /// User for primary and secondary progress.
  final Color? primaryColor;

  /// Secondary color that used last in gradient.
  /// User for primary and secondary progress.
  final Color? secondaryColor;

  /// Tertiary color that used for inactive part of progress.
  /// User for primary and secondary progress.
  final Color? tertiaryColor;

  @override
  void paint(Canvas canvas, Size size) {
    _drawPrimaryProgress(canvas, size);
    _drawSecondaryProgress(canvas, size);
  }

  void _drawPrimaryProgress(Canvas canvas, Size size) {
    if (primaryValue == null) return;

    final secondarySpace =
        secondaryValue != null ? secondaryWidth! + progressGap! : 0.0;
    final extraSpace = max(levelHighHeight!, levelLowHeight!) + secondarySpace;
    final activeRadius = radius! - extraSpace;
    final anglePerItem = maxDegrees! / levelAmount!;

    final paint = Paint()
      ..strokeWidth = levelLowWidth!
      ..strokeCap = StrokeCap.round
      ..style = PaintingStyle.stroke
      ..color = tertiaryColor!;

    for (var index = 0; index < levelAmount!; index++) {
      final angle = anglePerItem * index + startAngle! + (anglePerItem / 2);

      final isFillWithColor =
          (index / levelAmount!) <= primaryValue! && primaryValue != 0.0;

      final isHighLevel =
          division != null && index > 0 && index < levelAmount! - 1 ||
                  (levelHighBeginEnd! && index == 0) ||
                  (levelHighBeginEnd! && index == levelAmount! - 1)
              ? index % division! == 0
              : false;

      canvas.save();

      paint
        ..color = isFillWithColor
            ? ColorTween(
                begin: primaryColor,
                end: secondaryColor ?? primaryColor,
              ).transform(index / levelAmount!)!
            : tertiaryColor!
        ..strokeWidth = isHighLevel ? levelHighWidth! : levelLowWidth!;

      final offset = Offset(
        activeRadius * cos(pi * angle / 180) + activeRadius + extraSpace,
        activeRadius * sin(pi * angle / 180) + activeRadius + extraSpace,
      );

      canvas.translate(offset.dx, offset.dy);

      canvas.rotate(radians(angle));
      canvas.drawLine(
        Offset.zero,
        Offset(isHighLevel ? levelHighHeight! : levelLowHeight!, 0),
        paint,
      );

      canvas.restore();
    }
  }

  void _drawSecondaryProgress(Canvas canvas, Size size) {
    if (secondaryValue == null) return;

    final halfWidth = secondaryWidth! / 2;
    final secondRect = Rect.fromLTWH(
      halfWidth,
      halfWidth,
      size.width - secondaryWidth!,
      size.height - secondaryWidth!,
    );

    final secondPaint = Paint()
      ..style = PaintingStyle.stroke
      ..strokeWidth = secondaryWidth!
      ..strokeCap = StrokeCap.round
      ..color = tertiaryColor!;

    canvas.save();

    canvas.drawArc(
      secondRect,
      radians(startAngle!),
      radians(maxDegrees!),
      false,
      secondPaint,
    );

    secondPaint
      ..color = Colors.black
      ..shader = SweepGradient(
        colors: [
          primaryColor!,
          secondaryColor ?? primaryColor!,
        ],
        startAngle: radians(startAngle!),
        endAngle: radians(maxDegrees!),
        transform:
            GradientRotation(radians(startAngle! - (secondaryWidth! / 1.5))),
      ).createShader(secondRect);

    canvas.drawArc(
      secondRect,
      radians(startAngle!),
      radians(maxDegrees! * secondaryValue!),
      false,
      secondPaint,
    );

    canvas.restore();
  }

  @override
  bool shouldRepaint(covariant CustomPainter oldDelegate) {
    return this != oldDelegate;
  }
}


example/lib/pages/smart_oven.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:flutter_advanced_progress/flutter_advanced_progress.dart';

class SmartOverPage extends StatefulWidget {
  @override
  _SmartOverPageState createState() => _SmartOverPageState();
}

class _SmartOverPageState extends State<SmartOverPage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Container(
        padding: const EdgeInsets.symmetric(
          horizontal: 24,
        ),
        alignment: Alignment.center,
        child: Column(
          mainAxisSize: MainAxisSize.min,
          children: [
            Container(
              padding: const EdgeInsets.all(10),
              decoration: BoxDecoration(
                shape: BoxShape.circle,
                gradient: RadialGradient(
                  colors: [
                    Color.fromRGBO(128, 128, 138, 1),
                    Color.fromRGBO(53, 54, 62, 1),
                  ],
                  center: const Alignment(1, 1),
                  focal: const Alignment(-0.75, -0.75),
                  focalRadius: 1.0,
                ),
                boxShadow: [
                  const BoxShadow(
                    color: Colors.black38,
                    blurRadius: 16.0,
                    offset: const Offset(8.0, 8.0),
                  ),
                ],
              ),
              child: AdvancedProgress(
                radius: 120,
                levelAmount: 76,
                levelLowHeight: 16,
                levelHighHeight: 20,
                division: 15,
                secondaryWidth: 8,
                progressGap: 12,
                primaryValue: 0.50,
                secondaryValue: 0.75,
                primaryColor: Colors.yellow,
                secondaryColor: Colors.red,
                tertiaryColor: Colors.black12,
                child: Container(
                  alignment: Alignment.center,
                  child: Column(
                    mainAxisSize: MainAxisSize.max,
                    children: [
                      Spacer(),
                      Padding(
                        padding: const EdgeInsets.only(
                          top: 16.0,
                        ),
                        child: Text(
                          ' 240°',
                          textAlign: TextAlign.center,
                          style: TextStyle(
                            letterSpacing: 1.5,
                            fontFamily: 'Barlow',
                            fontWeight: FontWeight.w400,
                            fontSize: 50,
                            color: Colors.white,
                          ),
                        ),
                      ),
                      Padding(
                        padding: const EdgeInsets.only(
                          top: 16.0,
                        ),
                        child: Text(
                          'PREHEATING',
                          textAlign: TextAlign.center,
                          style: TextStyle(
                            letterSpacing: 3.0,
                            fontFamily: 'Barlow',
                            fontWeight: FontWeight.w800,
                            fontSize: 8,
                            color: Colors.deepOrange,
                          ),
                        ),
                      ),
                      Padding(
                        padding: const EdgeInsets.only(
                          top: 16.0,
                          bottom: 16.0,
                        ),
                        child: Column(
                          mainAxisSize: MainAxisSize.min,
                          children: [
                            Text(
                              '35:00',
                              textAlign: TextAlign.center,
                              style: TextStyle(
                                fontFamily: 'Barlow',
                                fontWeight: FontWeight.w400,
                                fontSize: 16,
                                color: Colors.white24,
                              ),
                            ),
                            Text(
                              'Time left',
                              textAlign: TextAlign.center,
                              style: TextStyle(
                                height: 1.5,
                                fontFamily: 'Barlow',
                                fontWeight: FontWeight.w700,
                                fontSize: 12,
                                color: Colors.white24,
                              ),
                            ),
                          ],
                        ),
                      ),
                    ],
                  ),
                ),
              ),
            ),
            Padding(
              padding: const EdgeInsets.only(
                left: 60,
                right: 60,
                top: 48,
              ),
              child: Container(
                height: 50,
                decoration: BoxDecoration(
                  color: Colors.black12,
                  borderRadius: BorderRadius.horizontal(
                    left: const Radius.circular(25),
                    right: const Radius.circular(25),
                  ),
                ),
                child: Row(
                  mainAxisSize: MainAxisSize.max,
                  children: [
                    Expanded(
                      child: Text(
                        'Swipe to Stop',
                        textAlign: TextAlign.center,
                        style: TextStyle(
                          fontFamily: 'Barlow',
                          fontWeight: FontWeight.w500,
                          fontSize: 16,
                        ),
                      ),
                    ),
                    Container(
                      width: 120,
                      height: double.maxFinite,
                      decoration: BoxDecoration(
                        color: Colors.redAccent,
                        borderRadius: BorderRadius.horizontal(
                          left: const Radius.circular(25),
                          right: const Radius.circular(25),
                        ),
                        gradient: LinearGradient(
                          colors: [
                            Colors.orange,
                            Colors.red,
                          ],
                        ),
                      ),
                      child: Icon(
                        Icons.chevron_left,
                        color: Colors.white,
                        size: 32,
                      ),
                    ),
                  ],
                ),
              ),
            ),
          ],
        ),
      ),
    );
  }
}


example/lib/pages/examples_page.dart
import 'package:flutter/material.dart';
import 'package:flutter_advanced_progress/flutter_advanced_progress.dart';

class ExamplesPage extends StatefulWidget {
  @override
  _ExamplesPageState createState() => _ExamplesPageState();
}

class _ExamplesPageState extends State<ExamplesPage>
    with TickerProviderStateMixin {
  late AnimationController _primaryAnimController;
  late AnimationController _secondaryAnimController;
  int _sliderAmount = 75;
  int _division = 15;

  @override
  void initState() {
    super.initState();

    _primaryAnimController = AnimationController(
      vsync: this,
      duration: Duration(milliseconds: 1000),
      value: 0.8,
    );

    _secondaryAnimController = AnimationController(
      vsync: this,
      duration: Duration(milliseconds: 500),
      value: 0.8,
    );
  }

  @override
  Widget build(BuildContext context) {
    final theme = Theme.of(context);

    return Scaffold(
      body: Container(
        padding: const EdgeInsets.all(8.0),
        alignment: Alignment.center,
        child: Column(
          mainAxisSize: MainAxisSize.min,
          children: [
            Row(
              mainAxisSize: MainAxisSize.max,
              children: [
                Container(
                  padding: const EdgeInsets.all(10),
                  decoration: BoxDecoration(
                    gradient: RadialGradient(
                      colors: [Colors.black87, Colors.black12],
                      center: Alignment.bottomCenter,
                      radius: 1.5,
                    ),
                    shape: BoxShape.circle,
                    boxShadow: [
                      BoxShadow(
                        color: Colors.black45,
                        blurRadius: 16.0,
                        offset: Offset(8.0, 8.0),
                      )
                    ],
                  ),
                  child: AnimatedBuilder(
                    animation: _primaryAnimController,
                    builder: (context, child) {
                      return AnimatedBuilder(
                        animation: _secondaryAnimController,
                        builder: (context, child) {
                          final celsius = (300 * _primaryAnimController.value)
                              .toStringAsFixed(0);

                          return AdvancedProgress(
                            levelAmount: _sliderAmount,
                            division: _division,
                            primaryValue: Tween<double>(
                              begin: 0.0,
                              end: 1.0,
                            ).evaluate(_primaryAnimController),
                            secondaryValue: Tween<double>(
                              begin: 0.0,
                              end: 1.0,
                            ).evaluate(_secondaryAnimController),
                            radius: 100.0,
                            levelLowHeight: 16,
                            levelHighHeight: 20,
                            primaryColor: Colors.yellow,
                            secondaryColor: Colors.red,
                            tertiaryColor: Colors.grey.withOpacity(0.25),
                            child: Stack(
                              children: [
                                Center(
                                  child: Column(
                                    mainAxisSize: MainAxisSize.min,
                                    children: [
                                      Text(
                                        '$celsius°',
                                        style: TextStyle(
                                          color: Colors.white,
                                          fontWeight: FontWeight.w500,
                                          fontSize: 36,
                                        ),
                                      ),
                                      Text(
                                        'PREHEATING',
                                        style: TextStyle(
                                          height: 2,
                                          color: Colors.red,
                                          fontWeight: FontWeight.w500,
                                          fontSize: 10,
                                        ),
                                      ),
                                    ],
                                  ),
                                ),
                                Align(
                                  alignment: Alignment.bottomCenter,
                                  child: Column(
                                    mainAxisSize: MainAxisSize.min,
                                    children: [
                                      Text(
                                        '35:00',
                                        style: TextStyle(
                                          fontSize: 14,
                                          fontWeight: FontWeight.w500,
                                          color: Colors.grey,
                                        ),
                                      ),
                                      Text(
                                        'Time left',
                                        style: TextStyle(
                                          fontSize: 12,
                                          fontWeight: FontWeight.w500,
                                          color: Colors.grey,
                                        ),
                                      ),
                                    ],
                                  ),
                                ),
                              ],
                            ),
                          );
                        },
                      );
                    },
                  ),
                ),
                Expanded(
                  child: Column(
                    mainAxisSize: MainAxisSize.min,
                    children: [
                      Slider(
                        value: _sliderAmount.toDouble(),
                        min: 0,
                        max: 200,
                        onChanged: (value) => setState(() {
                          _sliderAmount = value.toInt();
                        }),
                      ),
                      AnimatedBuilder(
                        animation: _primaryAnimController,
                        builder: (context, child) {
                          return Slider(
                            value: _primaryAnimController.value,
                            onChanged: (value) => setState(() {
                              _primaryAnimController.value = value;
                            }),
                          );
                        },
                      ),
                      AnimatedBuilder(
                        animation: _secondaryAnimController,
                        builder: (context, child) {
                          return Slider(
                            value: _secondaryAnimController.value,
                            onChanged: (value) => setState(() {
                              _secondaryAnimController.value = value;
                            }),
                          );
                        },
                      ),
                      Padding(
                        padding: const EdgeInsets.only(
                          top: 24,
                        ),
                        child: ElevatedButton(
                          onPressed: _handleDemoPressed,
                          child: Text('Demo'),
                        ),
                      ),
                    ],
                  ),
                ),
              ],
            ),
            SizedBox(
              height: 16.0,
            ),
            Row(
              mainAxisSize: MainAxisSize.max,
              mainAxisAlignment: MainAxisAlignment.spaceEvenly,
              children: [
                AdvancedProgress(
                  division: 5,
                  primaryValue: 0.75,
                  radius: 50.0,
                  levelAmount: 50,
                  levelLowHeight: 16,
                  levelHighHeight: 20,
                  primaryColor: Colors.green,
                  secondaryColor: Colors.red,
                  tertiaryColor: theme.hintColor,
                ),
                AdvancedProgress(
                  secondaryValue: 0.75,
                  secondaryWidth: 10.0,
                  radius: 50.0,
                  primaryColor: Colors.green,
                  secondaryColor: Colors.red,
                  tertiaryColor: theme.hintColor,
                ),
                AdvancedProgress(
                  radius: 50.0,
                  division: 5,
                  primaryValue: 1.0,
                  secondaryValue: 1.0,
                  levelAmount: 50,
                  primaryColor: Colors.black,
                  secondaryColor: Colors.white,
                  tertiaryColor: theme.hintColor,
                ),
              ],
            ),
            SizedBox(
              height: 16.0,
            ),
            Row(
              mainAxisSize: MainAxisSize.max,
              mainAxisAlignment: MainAxisAlignment.spaceAround,
              children: [
                AdvancedProgress(
                  division: 5,
                  primaryValue: 0.75,
                  radius: 50.0,
                  startAngle: -270,
                  maxDegrees: 180,
                  levelAmount: 50,
                  levelLowHeight: 16,
                  levelHighHeight: 20,
                  primaryColor: Colors.green,
                  secondaryColor: Colors.red,
                  tertiaryColor: theme.hintColor,
                ),
                AdvancedProgress(
                  division: 5,
                  primaryValue: 0.75,
                  radius: 50.0,
                  startAngle: -90,
                  maxDegrees: 180,
                  levelAmount: 50,
                  levelLowHeight: 16,
                  levelHighHeight: 20,
                  primaryColor: Colors.green,
                  secondaryColor: Colors.red,
                  tertiaryColor: theme.hintColor,
                ),
                AdvancedProgress(
                  division: 5,
                  primaryValue: 0.75,
                  radius: 50.0,
                  startAngle: 270,
                  maxDegrees: -180,
                  levelAmount: 50,
                  levelLowHeight: 16,
                  levelHighHeight: 20,
                  primaryColor: Colors.green,
                  secondaryColor: Colors.red,
                  tertiaryColor: theme.hintColor,
                ),
                AdvancedProgress(
                  division: 5,
                  primaryValue: 0.75,
                  radius: 50.0,
                  startAngle: 90,
                  maxDegrees: -180,
                  levelAmount: 50,
                  levelLowHeight: 16,
                  levelHighHeight: 20,
                  primaryColor: Colors.green,
                  secondaryColor: Colors.red,
                  tertiaryColor: theme.hintColor,
                ),
              ],
            ),
          ],
        ),
      ),
    );
  }

  void _handleDemoPressed() {
    _primaryAnimController.value = 0.0;
    _primaryAnimController.animateTo(
      0.8,
      duration: Duration(milliseconds: 500),
      curve: Curves.linearToEaseOut,
    );

    _secondaryAnimController.value = 0.0;
    _secondaryAnimController.animateTo(
      0.8,
      duration: Duration(milliseconds: 4000),
      curve: Curves.linearToEaseOut,
    );
  }
}


lib/src/widget.dart
part of '../flutter_advanced_progress.dart';

/// Advanced progress widget.
/// Represents two progress, primary and secondary.
class AdvancedProgress extends StatelessWidget {
  AdvancedProgress({
    Key? key,
    this.primaryValue,
    this.secondaryValue,
    this.secondaryWidth = 10.0,
    required this.radius,
    this.startAngle = 120.0,
    this.maxDegrees = 300.0,
    this.progressGap = 0.0,
    this.division = 10,
    this.levelAmount,
    this.levelLowWidth = 1.0,
    this.levelLowHeight = 8.0,
    this.levelHighWidth = 2.0,
    this.levelHighHeight = 16.0,
    this.levelHighBeginEnd = false,
    this.primaryColor,
    this.secondaryColor,
    this.tertiaryColor,
    this.child,
  }) : super(key: key);

  /// Value for primary progress.
  final double? primaryValue;

  /// Value for secondary progress.
  final double? secondaryValue;

  /// Secondary progress width.
  final double secondaryWidth;

  /// Total radius for whole widget.
  final double radius;

  /// Progress start angle.
  final double startAngle;

  /// Progress degrees from [startAngle].
  final double maxDegrees;

  /// Gap between primary and secondary progress.
  final double progressGap;

  /// Primary progress division.
  final int division;

  /// Amount of levels on primary progress.
  final int? levelAmount;

  /// Width of levels on primary progress.
  final double levelLowWidth;

  /// Height of low levels on primary progress.
  final double levelLowHeight;

  /// Height of high levels managed by [division] on primary progress.
  final double levelHighHeight;

  /// Width of levels on primary progress.
  final double levelHighWidth;

  /// True if need to begin and end with high level.
  final bool levelHighBeginEnd;

  /// Primary color that used as a color for progress of first in gradient.
  /// User for primary and secondary progress.
  final Color? primaryColor;

  /// Secondary color that used last in gradient.
  /// User for primary and secondary progress.
  final Color? secondaryColor;

  /// Tertiary color that used for inactive part of progress.
  /// User for primary and secondary progress.
  final Color? tertiaryColor;

  /// Child widget.
  final Widget? child;

  @override
  Widget build(BuildContext context) {
    final theme = Theme.of(context);

    final size = radius * 2;

    return Container(
      clipBehavior: Clip.antiAlias,
      width: size,
      height: size,
      decoration: BoxDecoration(),
      child: CustomPaint(
        painter: AdvancedProgressPainter(
          primaryValue: primaryValue,
          secondaryValue: secondaryValue,
          secondaryWidth: secondaryWidth,
          radius: radius,
          startAngle: startAngle,
          maxDegrees: maxDegrees,
          progressGap: progressGap,
          division: division,
          levelAmount: levelAmount,
          levelLowWidth: levelLowWidth,
          levelLowHeight: levelLowHeight,
          levelHighWidth: levelHighWidth,
          levelHighHeight: levelHighHeight,
          levelHighBeginEnd: levelHighBeginEnd,
          primaryColor: primaryColor ?? theme.indicatorColor,
          secondaryColor: secondaryColor ?? primaryColor,
          tertiaryColor: tertiaryColor ?? theme.hintColor,
        ),
        child: child,
      ),
    );
  }
}
